File: services/course.js
const uuid = require('uuid');
const db = require('../db/course');

const createCourse = course => {
  const c = course;
  c.id = uuid.v4();
  c.createdOn = new Date();
  c.lastEditedOn = new Date();
  return db.createCourse(c);
};

const getCourses = email => db.getCourses(email);

const getCourse = (id, email) => db.getCourse(id, email);

const updateCourse = course => {
  const c = course;
  c.lastEditedOn = new Date();
  return db.updateCourse(c);
};

const deleteCourse = (id, email) => db.deleteCourse(id, email);

module.exports = {
  createCourse,
  getCourses,
  getCourse,
  updateCourse,
  deleteCourse
};

File: pkg/controller/course.go
package controller

import (
	"context"
	"time"

	"github.com/ZdravkoGyurov/grader/pkg/types"
	"github.com/google/uuid"
)

func (c *Controller) CreateCourse(ctx context.Context, course *types.Course, userID, userName string) (*types.Course, error) {
	course.ID = uuid.NewString()
	course.CreatedOn = time.Now()
	course.LastEditedOn = time.Now()

	if err := course.ValidateCreate(); err != nil {
		return nil, err
	}

	courseGitlabID, err := c.createGitlabGroup(ctx, course.Name, course.GitlabName)
	if err != nil {
		return nil, err
	}
	course.GitlabID = courseGitlabID

	userProjectID, err := c.createGitlabProject(ctx, userName, courseGitlabID)
	if err != nil {
		return nil, err
	}

	if err := c.addUserInGitlabProject(ctx, userID, userProjectID); err != nil {
		return nil, err
	}

	if err := c.storage.CreateCourse(ctx, course); err != nil {
		return nil, err
	}

	return course, nil
}

func (c *Controller) GetCourses(ctx context.Context, userEmail string) ([]*types.Course, error) {
	return c.storage.GetCourses(ctx, userEmail)
}

func (c *Controller) GetCourse(ctx context.Context, id, userEmail string) (*types.Course, error) {
	return c.storage.GetCourse(ctx, id, userEmail)
}

func (c *Controller) UpdateCourse(ctx context.Context, course *types.Course) (*types.Course, error) {
	course.LastEditedOn = time.Now()

	if err := course.ValidateUpdate(); err != nil {
		return nil, err
	}

	return c.storage.UpdateCourse(ctx, course)
}

func (c *Controller) DeleteCourse(ctx context.Context, id, userEmail string) error {
	return c.storage.DeleteCourse(ctx, id, userEmail)
}
